pulp_smash/tests/rpm/cli/test_copy_units.py

```diff
@@ -181,3 +270,131 @@ def test_copy_langpacks(self):
                 _count_langpacks(self.cfg, self.repo_id),
                 0,
             )
+
+
+class CopyAndPublishTwoVersionsRepoTestCase(CopyBaseTestCase):
+    """Test whether a repo can copy two versions of RPMs and publish itself.
+
+    This test targets `Pulp Smash #311`_. The test steps are as follow:
+
+    1. Create a repo1 and sync from upstream.
+    2. Find a RPM that has two different versions in this repo.
+    3. Create a new repo2 and copy the older of the RPM version into it.
+    4. Publish the repo2 as a host.
+    5. Copy a newer version into the repo2 and re-publish.
+    6. Yum update on the client and verify update would succeed.
+
+    Note that the 1st repo `_REPO_ID` has been synced in `setUpModule()`,
+    while the 2nd repo `repo_id` has been created by base class without
+    a feed.
+
+    .. _Pulp Smash #311: https://github.com/PulpQE/pulp-smash/issues/311
+    """
+
+    @classmethod
+    def setUpClass(cls):
+        """Create two repositories and synchronize the 1st repo only."""
+        super(CopyAndPublishTwoVersionsRepoTestCase, cls).setUpClass()
+        # Retrieve all modules with multiple versions in the repo1.
+        cls.rpm_dict = {
+            key: value for (key, value)
+            in _get_rpm_names_versions(cls.cfg, _REPO_ID).items()
+            if len(value) > 1
+        }
+        assert len(cls.rpm_dict) > 0
+        # Choose the first module with multiple versions.
+        cls.rpm_name = cls.rpm_dict.keys()[0]
+        cls.rpm_dict.get(cls.rpm_name).sort()
+
+    def test_01_copy_older_publish(self):
+        """Copy a RPM with older version into repo2 and publish repo2."""
+        # Choose the oldest version of the given RPM.
+        rpm_version = self.rpm_dict.get(self.rpm_name, [None, None])[0]
+
+        # Copy dependency packages of the chosen RPM.
+        for rpm in _get_rpm_dependencies(self.cfg, _REPO_ID, self.rpm_name):
+            _copy_repo(self.cfg, _REPO_ID, self.repo_id, rpm)
+
+        self._copy_and_publish(self.rpm_name, rpm_version)
+        # Setup the repo2 on the same host.
+        cli.Client(self.cfg).run(
+            'sudo yum-config-manager --add-repo {0}{1}'
+            .format('https://dev/pulp/repos/', self.repo_id).split()
+        )
+        cli.Client(self.cfg).run('sudo yum repolist enabled'.split())
+        # Execute `yum install` to deploy the RPM on the host.
+        completed_proc = cli.Client(self.cfg).run(
+            'sudo yum install -y --nogpgcheck {}'.format(self.rpm_name).split()
```

Comment:
@Ichimonji10: Sorry again for the failure. After trying to reproduce it for multiple times, I finally found out that, the `yum install` of a RPM 'walrus-0.71' may **fail sometimes, while in other times, it will succeed**. And the output is due to this:
```
CompletedProcess(args=[u'sudo', u'yum', u'update', u'walrus'], returncode=1,
stdout=u'Last metadata expiration check: 0:00:05 ago on Wed Jun 29 21:22:24 2016.
No match for argument: walrus\n',
stderr=u"Yum command has been deprecated, redirecting to '/usr/bin/dnf update walrus'.\
See 'man dnf' and 'man yum2dnf' for more information.
To transfer transaction metadata from yum to DNF, run:
'dnf install python-dnf-plugins-extras-migrate && dnf-2 migrate'

Failed to synchronize cache for repo 'dev_pulp_repos_repo2_', disabling.
Package walrus not installed, cannot update it.
Error: No packages marked for upgrade.\n")
```

As u can see, is it sort of related with **"synchronize cache"** for repository that is already existing? So first, I'll add try..catch module to make my script more robust. And then I suppose to execute a `yum clean` in order to get rid of this "sometimes works; sometime fails" problem.